processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByMergingQueryRunnerV2.java

```diff
@@ -0,0 +1,661 @@
+    this.mergeBufferPool = mergeBufferPool;
+    this.spillMapper = spillMapper;
+  }
+
+  @Override
+  public Sequence<Row> run(final Query queryParam, final Map responseContext)
+  {
+    final GroupByQuery query = (GroupByQuery) queryParam;
+
+    // CTX_KEY_MERGE_RUNNERS_USING_CHAINED_EXECUTION is here because realtime servers use nested mergeRunners calls
+    // (one for the entire query and one for each sink). We only want the outer call to actually do merging with a
+    // merge buffer, otherwise the query will allocate too many merge buffers. This is potentially sub-optimal as it
+    // will involve materializing the results for each sink before starting to feed them into the outer merge buffer.
+    // I'm not sure of a better way to do this without tweaking how realtime servers do queries.
+    final boolean forceChainedExecution = query.getContextBoolean(
+        CTX_KEY_MERGE_RUNNERS_USING_CHAINED_EXECUTION,
+        false
+    );
+    final GroupByQuery queryForRunners = query.withOverriddenContext(
+        ImmutableMap.<String, Object>of(CTX_KEY_MERGE_RUNNERS_USING_CHAINED_EXECUTION, true)
+    );
+
+    if (BaseQuery.getContextBySegment(query, false) || forceChainedExecution) {
+      return new ChainedExecutionQueryRunner(exec, queryWatcher, queryables).run(query, responseContext);
+    }
+
+    final AggregatorFactory[] combiningAggregatorFactories = new AggregatorFactory[query.getAggregatorSpecs().size()];
+    for (int i = 0; i < query.getAggregatorSpecs().size(); i++) {
+      combiningAggregatorFactories[i] = query.getAggregatorSpecs().get(i).getCombiningFactory();
+    }
+
+    final GroupByMergingKeySerdeFactory keySerdeFactory = new GroupByMergingKeySerdeFactory(
+        query.getDimensions().size(),
+        config.getMaxMergingDictionarySize() / concurrencyHint
+    );
+    final GroupByMergingColumnSelectorFactory columnSelectorFactory = new GroupByMergingColumnSelectorFactory();
+
+    final File temporaryStorageDirectory = new File(
+        System.getProperty("java.io.tmpdir"),
+        String.format("druid-groupBy-%s_%s", UUID.randomUUID(), query.getId())
+    );
+
+    final LimitedTemporaryStorage temporaryStorage = new LimitedTemporaryStorage(
+        temporaryStorageDirectory,
+        config.getMaxOnDiskStorage()
+    );
+
+    // Figure out timeoutAt time now, so we can apply the timeout to both the mergeBufferPool.take and the actual
+    // query processing together.
+    final long startTime = System.currentTimeMillis();
+    final Number timeout = query.getContextValue(QueryContextKeys.TIMEOUT, null);
+    final long timeoutAt = timeout == null ? -1L : startTime + timeout.longValue();
+
+    final ResourceHolder<ByteBuffer> mergeBufferHolder;
+
+    try {
+      mergeBufferHolder = mergeBufferPool.take(timeoutAt > startTime ? timeoutAt - startTime : -1);
+    }
+    catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw Throwables.propagate(e);
```

Comment:
no, it doesn't need to be closed if it wasn't actually used to write files. but I think I'll add a close anyway just to avoid anyone wondering the same.